api/debug.go

```diff
@@ -43,11 +70,93 @@ func (d *DebugAPI) TraceTransaction(
 	hash gethCommon.Hash,
 	_ *tracers.TraceConfig,
 ) (json.RawMessage, error) {
-	res, err := d.tracer.GetTransaction(hash)
+	receipt, err := d.receipts.GetByTransactionID(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	tracerConfig := json.RawMessage(tracerConfig)
+	tracerCtx := &tracers.Context{
+		BlockHash:   receipt.BlockHash,
+		BlockNumber: receipt.BlockNumber,
+		TxIndex:     int(receipt.TransactionIndex),
+		TxHash:      receipt.TxHash,
+	}
+	tracer, err := tracers.DefaultDirectory.New(tracerName, tracerCtx, tracerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := d.blocks.GetByHeight(receipt.BlockNumber.Uint64())
+	if err != nil {
+		return nil, err
+	}
+	// We need to re-execute the given transaction and all the
+	// transactions that precede it in the same block, based on
+	// the previous block state, to generate the correct trace.
+	previousBlock, err := d.blocks.GetByHeight(block.Height - 1)
+	if err != nil {
+		return nil, err
+	}
+	previousBlockHeight := previousBlock.Height
+
+	client, err := d.client.GetClientForHeight(previousBlockHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	exeClient, ok := client.(*grpc.Client)
+	if !ok {
+		return nil, fmt.Errorf("could not convert to execution client")
+	}
+
+	ledger, err := evm.NewRemoteLedger(exeClient.ExecutionDataRPCClient(), previousBlockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not create remote ledger for height: %d, with: %w", previousBlockHeight, err)
+	}
+
+	blockExecutor, err := evm.NewBlockExecutor(
+		previousBlock,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
```

Comment:
we should extract all this logic into a tracer component. I don't like for API layer to have too much logic, it starts to get duplicated and it's harder to test in isolation. I suggest extracting the above into a tracer component in the services that just takes in a config of the tracer.

Reply:
at that point we should also write unit tests with just the tracer component